File: javascript/strain/strain.js
/* eslint-disable  no-restricted-syntax, no-multi-spaces */

const keepOrDiscard = (list, func, wanted) => {
  const results = [];
  for (const elem of list) {
    if (func(elem) === wanted) {
      results.push(elem);
    }
  }
  return results;
};

const keep    = (list, func) => keepOrDiscard(list, func, true);
const discard = (list, func) => keepOrDiscard(list, func, false);

module.exports = { keep, discard };

File: swift/strain/Sources/Strain/Strain.swift
extension Array {
    func keep(_ f: (Element) -> Bool) -> [Element] {
        var result = [Element]()
        for elem in self {
            if f(elem) {
                result.append(elem)
            }
        }
        return result
    }

    func discard(_ f: (Element) -> Bool) -> [Element] {
        return self.keep { !f($0) }
    }
}